flow/flow.go

```diff
@@ -0,0 +1,303 @@
+package flow
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/micro/go-micro/v2/broker"
+	"github.com/micro/go-micro/v2/client"
+	"github.com/micro/go-micro/v2/logger"
+)
+
+type Flow interface {
+	// Init flow with options
+	Init(...Option) error
+	// Get flow options
+	Options() Options
+	// Create step in specific flow
+	CreateStep(flow string, step *Step) error
+	// Delete step from specific flow
+	DeleteStep(flow string, step *Step) error
+	// Replace step in specific flow
+	ReplaceStep(flow string, oldstep *Step, newstep *Step) error
+	// Lookup specific flow
+	Lookup(flow string) ([]*Step, error)
+	// Execute specific flow and returns request id and error, optionally fills rsp
+	Execute(flow string, req interface{}, rsp interface{}, opts ...ExecuteOption) (string, error)
+	// Resume specific paused flow execution by request id
+	Resume(flow string, reqID string) error
+	// Pause specific flow execution by request id
+	Pause(flow string, reqID string) error
+	// Abort specific flow execution by request id
+	Abort(flow string, reqID string) error
+	// State show state specific flow execution by request id
+	State(flow string, reqID string) (string, error)
+	// Result get result of the flow step
+	Result(flow string, reqID string, step *Step) ([]byte, error)
+	// Stop executor and drain active workers
+	Stop() error
+}
+
+type Step struct {
+	// name of step
+	ID string
+	// Retry count for step
+	Retry int
+	// Timeout for step
+	Timeout int
+	// Step operation to execute
+	Operation Operation
+	// Which step use as input
+	Input string
+	// Where to place output
+	Output string
+	// Steps that are required for this step
+	After []string
```

Comment:
Should a Step know steps before and after or is this the point of the flow? Or is this provided by the flow once its registered a step? I would probably put []*Step if its going to stay so its a double linked list in some ways.

Reply:
Yes, each step may known after which step it must run. If you provide full flow you already have linked list of steps. But if you need to have ability to insert flow, you need to says before/after it must run.

Reply:
I can convert After/Before from []string to []*Step, but internally flow uses only ID of steps to create vertices and connect it